README.md
# SCRU128: Sortable, Clock and Random number-based Unique identifier

SCRU128 ID is yet another attempt to supersede [UUID] for the users who need
decentralized, globally unique time-ordered identifiers. SCRU128 is inspired by
[ULID] and [KSUID] and has the following features:

- 128-bit unsigned integer type
- Sortable by generation time (as integer and as text)
- 25-digit case-insensitive textual representation (Base36)
- 48-bit millisecond Unix timestamp that ensures useful life until year 10889
- Up to 281 trillion time-ordered but unpredictable unique IDs per millisecond
- 80-bit [three-layer randomness](#design-notes-three-layer-randomness) for
  global uniqueness

Examples in the 25-digit canonical textual representation:

```
0372hg16csmsm50l8dikcvukc
0372hg16csmsm50l8djl6xi25
0372hg16csmsm50l8dmgepzz1
0372hg16csmsm50l8doir3827
0372hg16cy3nowracls909wcd
0372hg16cy3nowraclvp355ce
0372hg16cy3nowraclxf2ctzh
0372hg16cy3nowraclyunyjke
```

[UUID]: https://en.wikipedia.org/wiki/Universally_unique_identifier
[ULID]: https://github.com/ulid/spec
[KSUID]: https://github.com/segmentio/ksuid

## Implementations

- [C/C++](https://github.com/scru128/c)
- [Go](https://github.com/scru128/go-scru128)
- [Java (with Kotlin and Android compatibility)](https://github.com/scru128/java)
- [JavaScript](https://github.com/scru128/javascript)
- [Python (and command-line tools)](https://github.com/scru128/python)
- [Rust](https://github.com/scru128/rust)
- [Swift](https://github.com/scru128/swift-scru128)

If you are interested in implementing SCRU128, see also [SCRU128 Generator
Tester](https://github.com/scru128/gen_test).

## Specification v2.1.1

A SCRU128 ID is a 128-bit unsigned integer consisting of four terms:

```
timestamp * 2^80 + counter_hi * 2^56 + counter_lo * 2^32 + entropy
```

Where:

- `timestamp` is a 48-bit Unix timestamp in milliseconds (i.e., milliseconds
  elapsed since 1970-01-01 00:00:00+00:00, ignoring leap seconds).
- `counter_hi` is a 24-bit randomly initialized counter that is incremented by
  one when `counter_lo` reaches its maximum value. `counter_hi` is reset to a
  random number when `timestamp` has moved forward by one second or more since
  the last renewal of `counter_hi`.
  - Note: `counter_hi` effectively works like an entropy component (rather than
    a counter) that is refreshed only once per second.
- `counter_lo` is a 24-bit randomly initialized counter that is incremented by
  one for each new ID generated within the same `timestamp`. `counter_lo` is
  reset to a random number whenever `timestamp` moves forward. When `counter_lo`
  reaches its maximum value, `counter_hi` is incremented and `counter_lo` is
  reset to zero.
- `entropy` is a 32-bit random number renewed for each new ID generated.

This definition is equivalent to allocating four unsigned integer fields to a
128-bit space according to the following layout:

| Bit numbers  | Field name | Size    | Data type        | Byte order |
| ------------ | ---------- | ------- | ---------------- | ---------- |
| Msb 0 - 47   | timestamp  | 48 bits | Unsigned integer | Big-endian |
| Msb 48 - 71  | counter_hi | 24 bits | Unsigned integer | Big-endian |
| Msb 72 - 95  | counter_lo | 24 bits | Unsigned integer | Big-endian |
| Msb 96 - 127 | entropy    | 32 bits | Unsigned integer | Big-endian |

Note that this specification does not specify a canonical bit layout of SCRU128
ID. An implementation may employ any binary form of a 128-bit unsigned integer
to represent a SCRU128 ID.

### Textual representation

A SCRU128 ID is encoded in a string using the _Base36_ encoding. The Base36
denotes a SCRU128 ID as a 128-bit unsigned integer in the radix of 36 using the
digits of `0-9a-z` (`0123456789abcdefghijklmnopqrstuvwxyz`), with leading zeros
added to form a 25-digit canonical representation. The following pseudo equation
illustrates the encoding algorithm:

```
1993501768880490086615869617690763354
    =  0  * 36^24 +  3  * 36^23 +  7  * 36^22 + ... + 27  * 36^2 + 29  * 36^1 + 22
    = '0' * 36^24 + '3' * 36^23 + '7' * 36^22 + ... + 'r' * 36^2 + 't' * 36^1 + 'm'
    = "0372ijojuxuhjsfkeryi2mrtm"
```

Although a 25-digit Base36 numeral can encode more than 128-bit information, any
numeral greater than `f5lxx1zz5pnorynqglhzmsp33` (`2^128 - 1`, the largest
128-bit unsigned integer) is not a valid SCRU128 ID.

For the sake of uniformity, an encoder should use lowercase letters in encoding
IDs. A decoder, on the other hand, must always ignore cases when interpreting or
lexicographically sorting encoded IDs.

The Base36 encoding shown above is available by default in several languages
(e.g., `BigInteger#toString(int radix)` and `BigInteger(String val, int radix)`
constructor in Java). Another easy way to implement it is by using 128-bit or
arbitrary-precision integer division and modulo operations. The following C code
illustrates a naive algorithm based on normal arrays and integers:

```c
const uint8_t id[16] = {1,   127, 239, 57, 194, 100, 27,  165,
                        106, 148, 131, 24, 136, 65,  224, 90};

// convert byte array into digit value array
uint8_t digit_values[25] = {0};
for (int i = 0; i < 16; i++) {
  unsigned int carry = id[i];
  for (int j = 24; j >= 0; j--) {
    carry += digit_values[j] * 256;
    digit_values[j] = carry % 36;
    carry = carry / 36;
  }
}

// convert digit value array into string
static const char digits[] = "0123456789abcdefghijklmnopqrstuvwxyz";
char text[26];
for (int i = 0; i < 25; i++) {
  text[i] = digits[digit_values[i]];
}
text[25] = '\0';
puts(text); // 0372ijojuxuhjsfkeryi2mrtm
```

See [the attached reference code] for a comprehensive example and test vectors.

[the attached reference code]: https://github.com/scru128/spec/blob/v2.1.1/base36_128.c

### Special-purpose IDs

The IDs with `timestamp` set at zero or `2^48 - 1` are reserved for special
purposes (e.g., use as dummy, error, or example values) and must not be used or
assigned as an identifier of anything.

### Considerations

#### Quality of random numbers

A generator should employ a cryptographically strong random or pseudorandom
number generator to generate unpredictable IDs.

#### Counter overflow handling

Counter overflow occurs at an extremely low probability when the randomly
initialized `counter_hi` and `counter_lo` do not provide sufficient space for
the IDs generated within a millisecond. The recommended approach to handle
counter overflow is to increment `timestamp` and continue in the following way:

1.  Increment `timestamp` by one.
2.  Reset `counter_hi` to zero.
3.  Reset `counter_lo` to a random number.

This approach is recommended over other options such as the "sleep till next
tick" approach because this technique allows the generation of monotonically
ordered IDs in a non-blocking manner. Raising an error on a counter overflow is
generally not recommended because a counter overflow is not a fault of users of
SCRU128.

This approach results in a greater `timestamp` value than the real-time clock.
Such a gap between `timestamp` and the wall clock should be handled as a small
clock rollback discussed below.

#### Clock rollback handling

A SCRU128 generator relies on a real-time clock to ensure the monotonic order of
generated IDs; therefore, it cannot guarantee monotonicity when the clock moves
back. When a generator detects a clock rollback by comparing the up-to-date
timestamp from the system clock and the one embedded in the last generated ID,
the recommended treatment is:

1.  If the rollback is small enough (e.g., a few seconds), treat the `timestamp`
    of the last generated ID as the up-to-date one, betting that the wall clock
    will catch up soon.
2.  Otherwise, reset `timestamp` to the wall clock and `counter_hi` and
    `counter_lo` to random numbers if the monotonic order of IDs is not
    critically important, or raise an error if it is.

This approach keeps the monotonic order of IDs when a clock rollback is small,
while it otherwise resets the generator and proceeds as if another new generator
were created to minimize the chance of collision.

#### Stateless variant

A generator may fill `counter_hi` and `counter_lo` with random numbers if it
generates IDs infrequently. Such a stateless implementation is acceptable,
though not recommended, because the outcome is not distinguishable from
compliant IDs.

### Design notes: three-layer randomness

SCRU128 utilizes timestamps and counters to ensure the uniqueness of IDs
generated by a single generator, whereas it relies on 80-bit entropy in the use
cases with distributed generators. SCRU128 fills the 80-bit field with a random
number when a new ID is infrequently (less than one ID per second) generated.
For the distributed high-load use cases, SCRU128 assigns different lifetimes to
the three entropy components to improve the collision resistance:

1.  24-bit `counter_hi`: reset to a random number every second
2.  24-bit `counter_lo`: reset to a random number every millisecond
3.  32-bit `entropy`: reset to a random number for every new ID generated

The longer lifetimes of `counter_hi` and `counter_lo` reduce the number of
random numbers consumed and accordingly reduce the probability of at least one
collision because, for a given length of random bits, the less the number of
dice throws, the lower the chance of collision.

In other words, generators are assigned to 24-bit `counter_hi` buckets every
second, and thus they will not collide with each other as long as their buckets
differ, even if each generates a bunch of IDs. 24-bit random numbers usually
collide if millions of instances are generated, but the one-second interval of
`counter_hi` renewals decreases the number of trials drastically. Nevertheless,
`counter_hi` is refreshed every second to prevent potential attackers from
exploiting this field as a generator's fingerprint.

Even within the same bucket, the generators will not collide as long as initial
`counter_lo` values are sufficiently distant from each other. Such a near match
probability, if tried only once a millisecond, is much lower than [the simple
birthday collision probability] calculated over all the IDs generated within a
millisecond. `entropy` provides additional protection in the extremely rare
cases where both `counter_hi` and `counter_lo` collide, but it is primarily
intended to ensure a certain level of unguessability of consecutive IDs
generated by a single generator.

[the simple birthday collision probability]: https://en.wikipedia.org/wiki/Birthday_problem

### License

This work is licensed under a [Creative Commons Attribution 4.0 International
(CC BY 4.0) License].

[Creative Commons Attribution 4.0 International (CC BY 4.0) License]: http://creativecommons.org/licenses/by/4.0/


CHANGELOG.md
# Changelog

## v2.1.1 - 2023-08-16

- Fixed typo and punctuations

## v2.1.0 - 2023-06-21

- Switched default letter case for generators from uppercase to lowercase
- Added optimized code examples of Base36 encoding and decoding

## v2.0.1 - 2022-07-31

- Added bit layout guideline
- Made minor improvements

## v2.0.0 - 2022-05-01

### Changed

- Textual representation: 26-digit Base32 -> 25-digit Base36
- Field structure: { `timestamp`: 44 bits, `counter`: 28 bits, `per_sec_random`:
  24 bits, `per_gen_random`: 32 bits } -> { `timestamp`: 48 bits, `counter_hi`:
  24 bits, `counter_lo`: 24 bits, `entropy`: 32 bits }
- Timestamp epoch: 2020-01-01 00:00:00.000 UTC -> 1970-01-01 00:00:00.000 UTC

### Added

- Counter overflow and clock rollback handling guidance

## v1.0.2 - 2022-01-02

- Applied CC BY 4.0 license

## v1.0.1 - 2021-11-22

- Fixed typo

## v1.0.0 - 2021-10-27

- Initial stable release


base36_128.c
/** base36_128.c - Base36 reference implementation for 128-bit data */

#include <assert.h>
#include <stdint.h>

/**
 * Converts a digit value array in `in_base` into that in `out_base`.
 *
 * This function converts digit value arrays in any pair of bases from 2 to 256,
 * while in this file it is used only to convert between Base36 and Base256
 * (i.e., bytes). Conversion of a digit value array from/to a string is taken
 * care of by `encode()` and `decode()` functions.
 *
 * This function includes naive code (turned on by the `USE_NAIVE_CODE` flag)
 * that focuses on the illustration of the algorithm and optimized code that
 * demonstrates typical optimization approaches.
 */
static int convert_base(const uint8_t *in, int in_len, int in_base,
                        uint8_t *out, int out_len, int out_base) {
  for (int i = 0; i < out_len; i++) {
    out[i] = 0;
  }

#ifdef USE_NAIVE_CODE
  for (int i = 0; i < in_len; i++) {
    // read one digit from `in` for each outer loop
    uint_fast16_t carry = in[i];

    // fill in `out` from right to left, while carrying up prior result to left
    for (int j = out_len - 1; j >= 0; j--) {
      carry += out[j] * in_base;
      out[j] = carry % out_base;
      carry = carry / out_base;
    }
    if (carry != 0) {
      return -1; // too small out_len
    }
  }
#else
  /*
   * The refined code below applies the following techniques to the naive code:
   *
   * - use larger `carry` and read multiple `in` digits for each outer loop
   * - break inner loop when `carry` is zero and none of remaining `out` digits
   *   has been updated from initial value (zero)
   */

  // determine number of `in` digits to read for each outer loop
  int word_len = 1;
  uint64_t word_base = in_base; // set to in_base ^ word_len
  while (word_base <= UINT64_MAX / ((uint64_t)in_base * out_base)) {
    word_len++;
    word_base *= in_base;
  }

  int out_used = out_len - 1; // storage to memorize range of `out` filled

  // iterate over `in` word by word, having `i` point to head of each word
  int i = in_len % word_len;
  if (i > 0) {
    i -= word_len;
  }
  for (; i < in_len; i += word_len) {
    // read multiple `in` digits for each outer loop
    uint64_t carry = 0;
    for (int j = i < 0 ? 0 : i; j < i + word_len; j++) {
      carry = carry * in_base + in[j];
    }

    for (int j = out_len - 1; j >= 0; j--) {
      carry += out[j] * word_base;
      out[j] = carry % out_base;
      carry = carry / out_base;

      // break inner loop when `carry` and remaining `out` digits are all zero
      if (carry == 0 && j <= out_used) {
        out_used = j;
        break;
      }
    }
    if (carry != 0) {
      return -1; // too small out_len
    }
  }
#endif /* #ifdef USE_NAIVE_CODE */

  return 0; // success
}

/**
 * Encodes a 128-bit byte array in a 25-digit Base36 string.
 *
 * @param bytes 16-byte byte array
 * @param out 26-byte string (25 digits and terminating NUL)
 */
void encode(const uint8_t *bytes, char *out) {
  // Base36 digit characters
  static const char DIGITS[] = "0123456789abcdefghijklmnopqrstuvwxyz";

  // convert byte array into digit value array
  uint8_t digit_values[25];
  int err = convert_base(bytes, 16, 256, digit_values, 25, 36);
  assert(err == 0);

  // convert digit value array into string
  for (int i = 0; i < 25; i++) {
    out[i] = DIGITS[digit_values[i]];
  }
  out[25] = '\0';
}

/**
 * Decodes a 128-bit byte array from a 25-digit Base36 string.
 *
 * This function assumes target environments where `char` is compatible with
 * ASCII.
 *
 * @param text 26-byte string (25 digits and terminating NUL)
 * @param out 16-byte byte array
 * @return zero on success or non-zero on failure
 */
int decode(const char *text, uint8_t *out) {
  // O(1) map from ASCII code points to Base36 digit values
  static const uint8_t DECODE_MAP[128] = {
      0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
      0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
      0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
      0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
      0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0xff, 0xff,
      0xff, 0xff, 0xff, 0xff, 0xff, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10,
      0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c,
      0x1d, 0x1e, 0x1f, 0x20, 0x21, 0x22, 0x23, 0xff, 0xff, 0xff, 0xff, 0xff,
      0xff, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14,
      0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 0x20,
      0x21, 0x22, 0x23, 0xff, 0xff, 0xff, 0xff, 0xff};

  // convert string into digit value array
  uint8_t digit_values[25];
  for (int i = 0; i < 25; i++) {
    unsigned char code = text[i];
    if (code > 127 || DECODE_MAP[code] == 0xff) {
      return -1; // invalid digit character
    }
    digit_values[i] = DECODE_MAP[code];
  }
  if (text[25] != '\0') {
    return -1; // invalid length
  }

  // convert digit value array into byte array
  int err = convert_base(digit_values, 25, 36, out, 16, 256);
  if (err != 0) {
    return -1; // out of 128-bit value range
  }

  return 0; // success
}

/** Executes the implementation against prepared test cases. */
static void test_positive_cases(void) {
  struct TestCase {
    uint8_t bytes[16];
    char text[26];
  };

  const struct TestCase test_vector[] = {
      {{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
       "0000000000000000000000000"},
      {{0x01, 0x7f, 0xee, 0x7f, 0xef, 0x41, 0x7e, 0x2b, 0x34, 0x32, 0xac, 0x2e,
        0xc5, 0x53, 0x68, 0x7c},
       "0372hg16csmsm50l8dikcvukc"},
      {{0x01, 0x7f, 0xee, 0x7f, 0xef, 0x42, 0x7e, 0x2b, 0x34, 0x6c, 0x0f, 0xf4,
        0x14, 0xbb, 0xcf, 0xfd},
       "0372hg16cy3nowracls909wcd"},
      {{0x01, 0x7f, 0xef, 0x39, 0xc2, 0x64, 0x1b, 0xa5, 0x6a, 0x94, 0x83, 0x18,
        0x88, 0x41, 0xe0, 0x5a},
       "0372ijojuxuhjsfkeryi2mrtm"},
      {{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
        0xff, 0xff, 0xff, 0xff},
       "f5lxx1zz5pnorynqglhzmsp33"}};
  const int N_CASES = 5;

  for (int i = 0; i < N_CASES; i++) {
    const struct TestCase *e = &test_vector[i];

    char out_text[26];
    encode(e->bytes, out_text);
    for (int j = 0; j < 26; j++) {
      assert(e->text[j] == out_text[j]);
    }

    uint8_t out_bytes[16];
    int err = decode(e->text, out_bytes);
    assert(err == 0);
    for (int j = 0; j < 16; j++) {
      assert(e->bytes[j] == out_bytes[j]);
    }
  }
}

/** Executes the implementation against test cases that return error. */
static void test_negative_cases(void) {
  uint8_t out_bytes[16];
  int err;
  err = decode("0", out_bytes);
  assert(err != 0);
  err = decode("00000000000000000000000000", out_bytes);
  assert(err != 0);
  err = decode("f5lxx1zz5pn+rynqglhzmsp33", out_bytes);
  assert(err != 0);
  err = decode("f5lxx1zz5pnorynqglhzmsp34", out_bytes);
  assert(err != 0);
  err = decode("zzzzzzzzzzzzzzzzzzzzzzzzz", out_bytes);
  assert(err != 0);
}

int main(void) {
  test_positive_cases();
  test_negative_cases();
  return 0;
}
